django/apps/registry.py

```diff
@@ -69,53 +70,58 @@ def populate(self, installed_apps=None):
         # populate() might be called by two threads in parallel on servers
         # that create threads before initializing the WSGI callable.
         with self._lock:
-            if self.ready:
-                return
-
-            # app_config should be pristine, otherwise the code below won't
-            # guarantee that the order matches the order in INSTALLED_APPS.
-            if self.app_configs:
-                raise RuntimeError("populate() isn't reentrant")
-
-            # Phase 1: initialize app configs and import app modules.
-            for entry in installed_apps:
-                if isinstance(entry, AppConfig):
-                    app_config = entry
-                else:
-                    app_config = AppConfig.create(entry)
-                if app_config.label in self.app_configs:
+            try:
+                if self.ready:
+                    return
+
+                # If populate is called concurrently, the code below won't
+                # guarantee that the app_config order matches the order in
+                # INSTALLED_APPS.
```

Comment:
I'm... not quite sure what I meant with this comment. Also the problem isn't concurrent calls, it's reentrant calls. The lock prevents concurrent calls when we're reaching this point.

Perhaps it should be rewritten as:

```
# Prevent reentrant calls to avoid running AppConfig.ready() methods twice.
```